internal/armada/server/lease.go

```diff
@@ -134,10 +135,39 @@ func (q *AggregatedQueueServer) ReturnLease(ctx context.Context, request *api.Re
 	if e := checkPermission(q.permissions, ctx, permissions.ExecuteJobs); e != nil {
 		return nil, e
 	}
-	_, err := q.jobRepository.ReturnLease(request.ClusterId, request.JobId)
+
+	// Check how many times the same job has been retried already
+	retries, err := q.jobRepository.GetNumberOfRetryAttempts(request.JobId)
+	if err != nil {
+		return nil, err
+	}
+
+	maxRetries := int(q.schedulingConfig.MaxRetries)
+	if retries >= maxRetries {
+		_, err := q.ReportDone(ctx, &api.IdList{Ids: []string{request.JobId}})
+		if err != nil {
+			return nil, err
+		}
+
+		failureReason := fmt.Sprintf("Exceeded maximum number of retries: %d", maxRetries)
+		err = q.reportFailure(request.JobId, request.ClusterId, failureReason)
```

Comment:
I think it would be better to do this before call to ReportDone. That way there is no chance of deleting job info without creating the event.